Detectors/TPC/reconstruction/src/TPCCATracking.cxx

```diff
@@ -0,0 +1,156 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See https://alice-o2.web.cern.ch/ for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file TPCCATracking.cxx
+/// \author David Rohr
+
+#include "TPCReconstruction/TPCCATracking.h"
+#ifdef HAVE_O2_TPCCA_TRACKING_LIB
+#include "TObject.h"
+#include "AliHLTTPCCAO2Interface.h"
+#else
+class AliHLTTPCCAO2Interface {}; // Dummy class such that the compiler can create a destructor for the unique_ptr
+class AliHLTTPCCAClusterData {}; // same
+#endif
+
+#include "DetectorsBase/Track.h"
+#include "FairLogger.h"
+#include "TClonesArray.h"
+#include "TPCBase/Mapper.h"
+#include "TPCBase/PadRegionInfo.h"
+#include "TPCBase/Sector.h"
+#include "TPCReconstruction/TrackTPC.h"
+#include "TPCSimulation/Cluster.h"
+#include "TPCBase/ParameterDetector.h"
+#include "TPCBase/ParameterGas.h"
+#include "TPCBase/ParameterElectronics.h"
+
+using namespace o2::TPC;
+
+TPCCATracking::TPCCATracking() : mTrackingCAO2Interface(), mClusterData_UPTR(), mClusterData(nullptr) {
+}
+
+TPCCATracking::~TPCCATracking() {
+  deinitialize();
+}
+
+int TPCCATracking::initialize(const char* options) {
+#ifndef HAVE_O2_TPCCA_TRACKING_LIB
+  LOG(FATAL) << "O2 compiled without TPC CA Tracking library, cannot run TPC CA Tracker\n";
+  return (1);
+#else
+  mTrackingCAO2Interface.reset(new AliHLTTPCCAO2Interface);
+  int retVal = mTrackingCAO2Interface->Initialize(options);
+  if (retVal) {
+    mTrackingCAO2Interface.reset();
+  } else {
+    mClusterData_UPTR.reset(new AliHLTTPCCAClusterData[36]);
+    mClusterData = mClusterData_UPTR.get();
+  }
+  return (retVal);
+#endif
+}
+
+void TPCCATracking::deinitialize() {
+  mTrackingCAO2Interface.reset();
+  mClusterData_UPTR.reset();
+  mClusterData = NULL;
+}
+
+int TPCCATracking::runTracking(const TClonesArray* inputClusters, std::vector<TrackTPC>* outputTracks) {
+  if (mTrackingCAO2Interface == nullptr) return (1);
+  int retVal = 0;
+
+#ifdef HAVE_O2_TPCCA_TRACKING_LIB
+  const static ParameterDetector &detParam = ParameterDetector::defaultInstance();
+  const static ParameterGas &gasParam = ParameterGas::defaultInstance();
+  const static ParameterElectronics &elParam = ParameterElectronics::defaultInstance();
+  
+  const AliHLTTPCGMMergedTrack* tracks;
+  int nTracks;
+  const unsigned int* trackClusterIDs;
+
+  int nClusters[36] = {};
+  for (Int_t icluster = 0; icluster < inputClusters->GetEntries(); ++icluster) {
+    Cluster& cluster = *static_cast<Cluster*>(inputClusters->At(icluster));
```

Comment:
this doesn't do any check if the object is of correct type, does it?

Reply:
This is probably alright for the moment since a dynamic cast in a loop will be quite expensive. This problem would be solved once we go away from TClonesArrays and have type-aware containers.